src/org/opensolaris/opengrok/search/context/OGKUnifiedHighlighter.java

```diff
@@ -0,0 +1,289 @@
+     * @param fields
+     * @param docIter
+     * @param cacheCharsThreshold
+     * @return a defined instance
+     * @throws IOException if an I/O error occurs
+     */
+    @Override
+    protected List<CharSequence[]> loadFieldValues(String[] fields,
+        DocIdSetIterator docIter, int cacheCharsThreshold) throws IOException {
+
+        List<CharSequence[]> docListOfFields = new ArrayList<>(
+            cacheCharsThreshold == 0 ? 1 : (int) Math.min(64, docIter.cost()));
+
+        int sumChars = 0;
+        do {
+            int docId = docIter.nextDoc();
+            if (docId == DocIdSetIterator.NO_MORE_DOCS) {
+                break;
+            }
+            Document doc = searcher.doc(docId);
+
+            String path = doc.get(QueryBuilder.PATH);
+            String storedU = doc.get(QueryBuilder.U);
+            String content = getRepoFileContent(path, storedU);
+
+            CharSequence[] seqs = new CharSequence[fields.length];
+            Arrays.fill(seqs, content);
+            docListOfFields.add(seqs);
+
+            if (content != null) {
+                sumChars += content.length();
+            }
+        } while (sumChars <= cacheCharsThreshold && cacheCharsThreshold != 0);
+
+        return docListOfFields;
+    }
+
+    /**
+     * Returns the value from the {@code super} implementation, with logging for
+     * ANALYSIS of any field but {@link QueryBuilder#FULL} or
+     * {@link QueryBuilder#REFS}.
+     * @return the value from the {@code super} implementation
+     */
+    @Override
+    protected OffsetSource getOptimizedOffsetSource(String field,
+        BytesRef[] terms, PhraseHelper phraseHelper,
+        CharacterRunAutomaton[] automata) {
+
+        OffsetSource res = super.getOptimizedOffsetSource(field,
+            terms, phraseHelper, automata);
+        if (res == OffsetSource.ANALYSIS) {
+            /**
+             *     Testing showed that UnifiedHighlighter falls back to
+             * ANALYSIS in the presence of multi-term queries (MTQs) such as
+             * prefixes and wildcards even for fields that are analyzed with
+             * POSTINGS -- i.e. with DOCS_AND_FREQS_AND_POSITIONS_AND_OFFSETS.
+             * This is despite UnifiedHighlighter seeming to indicate that
+             * postings should be sufficient in the comment for
+             * shouldHandleMultiTermQuery(String): "MTQ highlighting can be
+             * expensive, particularly when using offsets in postings."
+             *     DEFS are stored with term vectors to avoid this problem.
+             *     FULL should be approximately fine with re-analysis using an
+             * on-the-fly PlainAnalyzer.
+             *     REFS should be approximately fine with re-analysis using an
+             * on-the-fly PlainAnalyzer. It might not accord with the true
+             * language symbol tokenizer, but it should not be wildly
+             * divergent.
```

Comment:
can we inject the language symbol tokenizer somehow?

Reply:
(fwiw we already do that when doing analysis, it's a bit hacky since lucene defaulted to just one default analyzer ( see https://github.com/oracle/opengrok/blob/master/src/org/opensolaris/opengrok/analysis/plain/PlainAnalyzer.java#L98 ) )

Reply:
OK after studying the Lucene `UnifiedHighlighter` implementation, I pushed an arcane override of `getIndexAnalyzer()` when OpenGrok's own `highlightFieldsUnion()` is executing so that an OpenGrok `TYPE`-specific analyzer is returned when re-`ANALYSIS` is being done. (I also updated those block comments.)